### PR TITLE
fix(manifest): encode decimal partitions with declared width

### PR DESCRIPTION
**Pull request by fallintoplace**
## Summary

Decimal partition values now use the fixed byte width declared by partition-schema precision instead of a width inferred from the formatted value.

- keep decimal read scale separate from fixed encoding width;
- sign-extend negative values to the declared width;
- reject oversized values unless discarded bytes are redundant sign extension;
- apply the declared width in data-file statistics conversion and `MarshalAvroEntry`;
- keep builder-created partition values directly typed; scale reconstruction is only required after Avro decoding.

## Testing

- table-driven positive, zero, negative, boundary, padding, and overflow vectors
- manifest write/read decimal round trips
- `DataFileStatistics.ToDataFile` → manifest round trip
- `MarshalAvroEntry` → decoder round trip
- `go test . ./table/internal`

### CODE REVIEW

File: manifest.go [hunk 1]
```diff
@@ -1848,18 +1864,28 @@ func convertTimestampMicrosValue(v any) any {
...
+		return nil, fmt.Errorf("missing decimal fixed size")
 	}
 
-	return v
+	bytes, err := dec.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+
+	return padOrTruncateBytes(bytes, fixedSize), nil
```
Comment (zeroshade) [inline manifest.go]: [BLOCKER] This call still routes through `padOrTruncateBytes`, whose implementation at lines 1899-1905 left-pads short decimal byte slices with `0x00` unconditionally. That is only correct for non-negative values; a negative unscaled value such as -1 for precision 10 (5 bytes) becomes `00 00 00 00 ff` and decodes as +255. Fix with sign-aware fixed-width encoding: when `len(bytes) < fixedSize`, prepend `0xff` if `bytes[0]&0x80 != 0`, otherwise `0x00` (two's-complement sign extension), and add negative/zero/max-precision-38 vectors.

Comment (zeroshade) [inline manifest.go]: [MAJOR] `padOrTruncateBytes` also silently truncates oversized decimals to the low-order bytes (see real lines 1899-1901), which can change the value instead of reporting that it does not fit the declared fixed size. Since this path already returns an error, make `len(bytes) > fixedSize` verify the dropped prefix is redundant sign extension and otherwise return an overflow/precision error.

File: manifest_test.go [hunk 2]
```diff
@@ -2044,6 +2046,25 @@ func TestReadManifestDecodesNilLogicalPartitionValueFromNullableUnion(t *testing
 	}
 }
 
+func TestAvroEncodePartitionDataUsesDeclaredDecimalFixedSize(t *testing.T) {
```
Comment (zeroshade) [inline manifest_test.go:2177]: [MINOR] This test only covers one positive decimal, so it misses the negative sign-extension bug and oversized-value truncation. Please make this table-driven with positive, zero, negative, max-width, and too-large vectors, and include a `ManifestWriter`/`ReadManifest` round-trip.

**Review by zeroshade — CHANGES_REQUESTED** [review]
Decimal partition encoding still corrupts negative/oversized fixed decimals and needs broader coverage.

- [BLOCKER] Fixed-width decimal padding must sign-extend negatives instead of always prepending 0x00.
- [MAJOR] Oversized decimals must be rejected unless the dropped prefix is redundant sign extension.
- [MINOR] Add table-driven decimal vectors and a manifest round-trip to cover this behavior.

File: manifest.go [hunk 3]
```diff
@@ -2006,7 +2061,7 @@ func (d *dataFile) convertAvroValueToIcebergType(v any, fieldID int) any {
 			return TimestampNano(v.(int64))
 		case atype.Decimal:
 			if r, ok := v.(*big.Rat); ok {
-				scale := d.fieldIDToFixedSize[fieldID]
+				scale := d.fieldIDToDecimalScale[fieldID]
```
Comment (laskoviymishka) [inline manifest.go:2122]: Now that the read path reads `d.fieldIDToDecimalScale` here, `d.fieldIDToFixedSize` has no reader left anywhere — the write path encodes off `ManifestWriter.partFieldIDToSize` (schema-derived), not the per-`dataFile` field. That makes the `fieldIDToFixedSize` argument to `NewDataFileBuilder` silently a no-op: a caller passing 0 or a wrong width gets neither an error nor a different result.

I'd drop `fieldIDToFixedSize` from the struct and the `NewDataFileBuilder` signature; if it has to stay for the `hasFieldToIDMap` interface, at least comment that it's unread on the write path so we're not handing callers a knob that does nothing.

File: table/internal/utils.go [hunk 4]
```diff
@@ -290,7 +291,7 @@ func (d *DataFileStatistics) ToDataFile(opts DataFileOpts) iceberg.DataFile {
 					fieldIDToLogicalType[field.FieldID] = atype.TimestampMicros
 				case iceberg.DecimalType:
 					fieldIDToLogicalType[field.FieldID] = atype.Decimal
-					fieldIDToFixedSize[field.FieldID] = rt.Scale()
+					fieldIDToFixedSize[field.FieldID] = iceinternal.DecimalRequiredBytes(rt.Precision())
```
Comment (laskoviymishka) [inline table/internal/utils.go]: This is the same class of bug the manifest fix addresses, but on the path Arrow writers actually use to build `DataFile`s for commit — and it has no test. A regression here would corrupt every decimal partition value written by an Arrow writer without a single test failing.

I'd add one that runs a `DataFileStatistics` with a decimal partition column through `ToDataFile` → `WriteManifest` → read back and asserts the decoded value round-trips.

File: manifest.go [hunk 5]
```diff
@@ -1916,6 +1970,7 @@ type dataFile struct {
 	fieldIDToLogicalType   map[int]string
 	fieldIDToPartitionData map[int]any
 	fieldIDToFixedSize     map[int]int
+	fieldIDToDecimalScale  map[int]int
```
Comment (laskoviymishka) [inline manifest.go:2031]: `fieldIDToDecimalScale` only ever gets populated by the `ReadEntry` setter path — `NewDataFileBuilder` doesn't take a scale map. So a `dataFile` built via the constructor and read back with `Partition()` without going through `ManifestReader` reads `scale=0`, and `(123, scale=2)` silently decodes as `123`.

Mirror image of the dead `fieldIDToFixedSize` above — I'd either add a `fieldIDToDecimalScale` param to `NewDataFileBuilder` or derive scale from the spec+schema in the constructor. wdyt?

File: manifest.go [hunk 6]
```diff
@@ -1870,13 +1896,41 @@ func convertUUIDValue(v any) any {
 	return v
 }
 
-func padOrTruncateBytes(bytes []byte, size int) []byte {
-	if len(bytes) >= size {
-		return bytes[len(bytes)-size:]
+func fitDecimalBytes(bytes []byte, size int) ([]byte, error) {
+	if len(bytes) == 0 {
```
Comment (laskoviymishka) [inline manifest.go]: The `len(bytes) == 0` branch is unreachable — `DecimalLiteral.MarshalBinary` always emits at least one byte (zero gives `[0x00]`), so no caller can deliver an empty slice. Meanwhile there's no `size <= 0` guard: called with `size == 0` and non-empty bytes, it falls through to the trim branch and `retained[0]` panics on the empty slice.

`encodeDecimalBytes` catches `fixedSize <= 0` today so it's not reachable in production, but since this is unexported and test-callable I'd swap the dead empty-bytes check for a `size <= 0` guard that returns an error.

File: manifest_test.go [hunk 7]
```diff
@@ -2044,6 +2047,202 @@ func TestReadManifestDecodesNilLogicalPartitionValueFromNullableUnion(t *testing
...
+				t.Fatal(err)
+			}
+			if !bytes.Equal(got, tt.want) {
+				t.Fatalf("fitDecimalBytes() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestAvroEncodePartitionDataUsesDeclaredDecimalFixedSize(t *testing.T) {
```
Comment (laskoviymishka) [inline manifest_test.go:2177]: The vectors here are solid, but they only hit precision 10 and 38 — I'd add the `requiredLength` tier boundaries too, since those are exactly where an off-by-one in `DecimalRequiredBytes` would show up: precision 1 (size 1, value 5 → `[0x05]`), precision 9 (size 4), and a negative at precision 18 (size 9, sign-extended).

While we're here, `TestFitDecimalBytesAllowsOnlyRedundantSignExtension` only exercises the trim path — a direct padding case like `{0x01}` size 3 → `{0x00,0x00,0x01}` plus a negative `{0xff}` → `{0xff,0xff,0xff}` would cover the pad branch directly rather than only through the encode test.

File: manifest.go [hunk 8]
```diff
@@ -376,7 +376,7 @@ func (m *manifestFile) FetchEntries(fs iceio.IO, discardDeleted bool) (_ []Manif
 	return ReadManifest(m, f, discardDeleted)
 }
 
-func getFieldIDMap(sc *avro.Schema) (map[string]int, map[int]string, map[int]int) {
+func getFieldIDMap(sc *avro.Schema) (map[string]int, map[int]string, map[int]int, map[int]int) {
```
Comment (laskoviymishka) [inline manifest.go]: `getFieldIDMap` now returns four positional maps, two of them `map[int]int` — at the call sites the size map and the scale map are indistinguishable by type, and one site already discards the fourth with `_`. I'd return the existing `dataFileFieldMaps` struct (or named returns) so a future caller can't swap size and scale by accident. wdyt?

File: manifest.go [hunk 9]
```diff
@@ -1848,18 +1864,28 @@ func convertTimestampMicrosValue(v any) any {
...
 }
 
-func convertDecimalValue(v any) any {
-	if dec, ok := v.(Decimal); ok {
-		fixedSize := internal.DecimalRequiredBytes(len(dec.String()))
-		bytes, err := DecimalLiteral(dec).MarshalBinary()
-		if err != nil {
-			return v
-		}
+func convertDecimalValue(v any, fixedSize int) (any, error) {
```
Comment (laskoviymishka) [inline manifest.go:1926]: `convertDecimalValue` now handles both `Decimal` (from `NewDataFileBuilder`) and `DecimalLiteral` (from a round-trip through `initPartitionData`) — that dual case is a bit of a symptom of the partition value model not converging on one type. Not blocking, but a one-line comment on when each shows up would save the next reader the archaeology. wdyt?

File: manifest.go [hunk 10]
```diff
@@ -1870,13 +1896,41 @@ func convertUUIDValue(v any) any {
...
 
-	return append(make([]byte, size-len(bytes)), padded...)
+	retained := bytes[len(bytes)-size:]
+	signByte := byte(0x00)
+	if retained[0]&0x80 != 0 {
+		signByte = 0xff
+	}
+	for _, b := range bytes[:len(bytes)-size] {
+		if b != signByte {
+			return nil, fmt.Errorf("decimal value does not fit fixed size %d", size)
```
Comment (laskoviymishka) [inline manifest.go]: Small thing: this reports only `size`, so at the top level you get `decimal value does not fit fixed size 5` with no hint whether the value was 6 bytes or 16. Including `len(bytes)` — `decimal value of %d bytes does not fit fixed size %d` — makes it debuggable at a glance.

File: manifest_test.go [hunk 11]
```diff
@@ -2044,6 +2047,202 @@ func TestReadManifestDecodesNilLogicalPartitionValueFromNullableUnion(t *testing
...
+			value:     Decimal{Val: decimal128.FromI64(-1), Scale: 2},
+			precision: 10,
+			want:      []byte{0xff, 0xff, 0xff, 0xff, 0xff},
+		},
+		{
+			name:      "keeps max precision decimal",
+			value:     DecimalLiteral{Val: decimal128.FromBigInt(maxPrecision38), Scale: 0},
+			precision: 38,
+			want: []byte{
+				0x4b, 0x3b, 0x4c, 0xa8, 0x5a, 0x86, 0xc4, 0x7a,
```
Comment (laskoviymishka) [inline manifest_test.go:2230]: Could we drop a one-liner on what these bytes are — e.g. `// two's-complement big-endian of 10^38 - 1`? A bare 16-byte literal is the kind of thing that's impossible to verify or update later without recomputing it by hand.

**Review by laskoviymishka — CHANGES_REQUESTED** [review]
Nice work on the revision — the pieces zeroshade asked for all look like they landed: negatives sign-extend in `fitDecimalBytes`, oversized values are rejected unless the dropped prefix is redundant sign extension, and there's a table-driven vector set plus a negative round-trip through `WriteManifest`/`ReadManifest`. The core fix is right, too — splitting write-width from read-scale in `getFieldIDMap` is exactly the bug, and the encoded bytes line up with Java and PyIceberg.

I'd hold before merging though, mostly on one thing I kept running into from different directions. Now that the read path reads `fieldIDToDecimalScale`, the old `fieldIDToFixedSize` state has no reader left — the write path encodes off the schema-derived `partFieldIDToSize`, not the per-`dataFile` field — but `NewDataFileBuilder` still accepts `fieldIDToFixedSize` as a parameter. So it's a public argument that silently does nothing, and its mirror (`fieldIDToDecimalScale`) has no constructor entry point at all, which means a builder-built file read back outside `ManifestReader` decodes at `scale=0`. I'd like the constructor surface to tell the truth about which of these actually affects the bytes.

The other one is the `DataFileStatistics.ToDataFile` change — same fix, right change, but it's the path Arrow writers use to produce data files for commit and it has no test, so a future regression there would corrupt decimal partitions with nothing failing.

Things I'd like to settle before merge:
- Make `NewDataFileBuilder` honest about `fieldIDToFixedSize` (drop it, or document it's unread on the write path) and give it a way to set scale.
- Add a test for the decimal path through `ToDataFile` → `WriteManifest` → read back.
- Extend the vectors to the `requiredLength` tier boundaries (precision 1/9/18), and cover `fitDecimalBytes`'s padding branch directly.

Everything else is small — a dead `len(bytes)==0` guard, error messages that could carry the byte length, the four positional maps out of `getFieldIDMap`. Fix the API surface and the `ToDataFile` test and I'm happy to take another pass and approve.

File: manifest.go [hunk 12]
```diff
@@ -2260,6 +2320,7 @@ func NewDataFileBuilder(
 	recordCount int64,
 	fileSize int64,
 ) (*DataFileBuilder, error) {
+	_ = fieldIDToFixedSize
```
Comment (laskoviymishka) [inline manifest.go:2387]: I said document-or-drop last round and you documented it — that's fair by what I asked. Seeing `_ = fieldIDToFixedSize` actually land though, I've come around to preferring drop.

Both reviewers on this pass read it as a live parameter and assumed it mattered, and the round-trip test still passes a real `internal.DecimalRequiredBytes(10)` into it that goes nowhere. We're pre-1.0, so dropping it is cheaper than carrying a dead arg everyone has to be told to ignore.

Not blocking — your call. Just flagging that's where I'd land now.

File: manifest.go [hunk 13]
```diff
@@ -2313,7 +2374,6 @@ func NewDataFileBuilder(
 			fieldIDToPartitionData: fieldIDToPartitionData,
 			fieldNameToID:          fieldNameToID,
 			fieldIDToLogicalType:   fieldIDToLogicalType,
```
Comment (laskoviymishka) [inline manifest.go:2440]: This is the scale half of what I asked last round and it's still open — a dataFile built here has no way to populate `fieldIDToDecimalScale`; only `ManifestReader.ReadEntry` sets it via `setFieldIDToDecimalScaleMap`. So a builder-built file decoded outside a manifest read hits `convertAvroValueToIcebergType` with scale=0 and mis-scales the decimal.

What I can't tell is whether that path is actually reachable — is anything meant to decode a builder-built dataFile without going through `ManifestReader`? If not, this is moot and I'll drop it; if it is, we'd want a scale setter on the builder. wdyt?

File: data_file_codec.go [hunk 14]
```diff
@@ -107,7 +107,11 @@ func (d *dataFile) MarshalAvroEntry(spec PartitionSpec, schema *Schema, version
 		return nil, err
 	}
 	clone := cloneDataFileAvroFields(d)
-	clone.PartitionData = avroEncodePartitionData(d.Partition(), maps.nameToID, maps.idToType)
+	partitionData, err := avroEncodePartitionData(d.Partition(), maps.nameToID, maps.idToType, maps.idToFixedSize)
```
Comment (laskoviymishka) [inline data_file_codec.go:110]: There are two write paths that encode decimal partitions — `ManifestWriter.addEntry` and this one — and the new tests only cover the first. The old bug lived in both and this fix touches both, so I'd want a round-trip here too: `MarshalAvroEntry` → `unmarshalAvroDataFileEntry` on a decimal partition. Otherwise a regression on this path slips through. Could you add one?

**Review by laskoviymishka — CHANGES_REQUESTED** [review]
Really nice revision. Everything I flagged last time is covered now.

`ToDataFile` has a round-trip test, the padding branch in `fitDecimalBytes` is covered directly, including the precision 1/9/18/38 boundaries, and `getFieldIDMap` now returns a `dataFileFieldMaps` struct instead of four positional maps. The encoding still matches Java and PyIceberg.

I have two narrow things left before approving.

First, the scale part of the builder question is still open. A `dataFile` created through `NewDataFileBuilder` has no way to populate `fieldIDToDecimalScale`; only `ManifestReader` does that. So if that file is decoded anywhere outside a manifest read, decimals would be interpreted with scale 0.

I’m not sure that path is actually reachable. If builder-created files are never decoded without going through `ManifestReader`, then this is moot and I’m happy to drop it. I’d just like to confirm that rather than assume.

Second, `MarshalAvroEntry` is another path that writes decimal partitions, and there still isn’t a decimal test for it. Since this change affects that path too, I’d like to see a round-trip test there.

On `fieldIDToFixedSize`, I asked for either documenting or removing it last time, and you documented it, which is fair. But now that `_ = fieldIDToFixedSize` is in the code, I think removing it would be cleaner. Both reviewers read it as live state, and the test passes a real value that is then ignored. Since we’re pre-1.0, this seems cheap to clean up. Not blocking, just my preference.

Everything else @zeroshade and I raised is in:

* negative values sign-extend correctly
* oversized values are rejected unless the removed prefix is redundant
* the table-driven vectors are there
* write width and read scale are separated cleanly
* the `ToDataFile` test and boundary cases landed

The smaller items from last round, like the `len(bytes) == 0` guard and including the byte length in errors, are still optional.

Once the scale question is settled and there’s a `MarshalAvroEntry` decimal test, I’m happy to approve.

**Review by zeroshade — APPROVED** [review]
Re-reviewed the current revision — my earlier blocker is resolved. `fitDecimalBytes` now sign-extends negative decimals (0xff when the high bit is set) and rejects oversized values unless the dropped prefix bytes are redundant sign extension, with the added round-trip/precision coverage. Thanks @fallintoplace! (Leaving @laskoviymishka's review to them.)

**Comment by zeroshade** [comment]
@laskoviymishka when you get a chance, could you take another look here? Your last note said everything you'd flagged was covered, and I re-reviewed the current revision — `fitDecimalBytes` now sign-extends negative decimals and rejects oversized values unless the dropped prefix is redundant sign extension, with the added round-trip/precision coverage. I've approved on my side; if you're happy with it, converting your review would let it merge. Thanks!

File: manifest.go [hunk 15]
```diff
@@ -2324,6 +2384,7 @@ func NewDataFileBuilder(
 	recordCount int64,
 	fileSize int64,
 ) (*DataFileBuilder, error) {
+	_ = fieldIDToFixedSize
```
Comment (zeroshade) [inline manifest.go:2387]: Non-blocking: `NewDataFileBuilder` still retains and ignores `fieldIDToFixedSize` (`_ = fieldIDToFixedSize`). Consider dropping the dead parameter now that the write path derives width from the manifest schema.

**Review by zeroshade — COMMENTED** [review]
Re-reviewed at head — the blocking decimal issues are resolved (negative sign-extension, error-on-oversized, both write paths + the `table/internal` Arrow path, and builder scale via the Avro decode). One non-blocking nit below. Note: #1509 targets the same bug; this PR supersedes it. CI green.

**Review by laskoviymishka — APPROVED** [review]
